<img src="logo_ESPRIT.jpg" width="300" align="right"></a>
<img src="breast.jpg" width="130" align="left"></a>

<h1 align="center"><font size="7">Projet Machine Learning 4CINFOGL1</font></h1>

Ce Porjet doit être rédigé et sera évalué. Le code python produit doit être inclus. Pour la mise en page et le format, il est préférable de le rendre sous la forme d’un notebook Jupyter, ou alors d’un document pdf. Les noms doivent apparaître dans le document. Il doit être envoyé le 30 Janvier 2021 au plus tard à l’adresse suivante.
lotfi.ncib@esprit.tn 

# Objectif du projet
L'objectif de ce projet est de:
- Tester cinq algorithmes de classification (K_nearst neighbors, Arbre de décision, Régression Logistique, Gaussian Naive Bayes et Support vector machine) pour la résolution d'un problème de classification binaire(deux classes)
- Régler le maximum de paramètres pour chaque algorithme
- Tracer la matrice de confusion et afficher __classification_report__ de chaque algorithme
- Choisir le meilleur algorithme en utilisant __classification_report__

# Base de données __breast_cancer__

Pour ce projet on considère un jeu de données médicales. Chaque exemple correspond à une tumeur du sein. Les variables explicatives portent sur des caractéristiques observées de la tumeur. La variable à prédire indique s’il s’agit d’une tumeur maligne (4) ou bénigne (2). 

In [2]:
# Chargement des bibliothèques
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Importer l'ensemble de données avec read_csv:  Variables __breast_cancer

In [4]:
data_breast_cancer= pd.read_csv('variables_breast_cancer.csv')
data_breast_cancer

name     type  \
0              clump_thickness  integer   
1      uniformity_of_cell_size  integer   
2     uniformity_of_cell_shape  integer   
3            marginal_adhesion  integer   
4  single_epithelial_cell_size  integer   
5                  bare_nuclei  integer   
6              bland_chromatin  integer   
7              normal_nucleoli  integer   
8                      mitosis  integer   
9                        class  integer   

                                      description  
0                               Value range: 1-10  
1                               Value range: 1-10  
2                               Value range: 1-10  
3                               Value range: 1-10  
4                               Value range: 1-10  
5                               Value range: 1-10  
6                               Value range: 1-10  
7                               Value range: 1-10  
8                               Value range: 1-10  
9  Predictor Value: 2 for benign, 4 for malignant

#### Question 1: Créer une variable `data` à qui vous affectez la base de données `breast_cancer.csv`

In [5]:
data=pd.read_csv('breast_cancer.csv')

#### Question 2: Quelle est la dimension de `data`?

In [6]:
data.shape

(569, 10)

#### Question 3:  En utilisant la méthode `head` (resp la méthode `tail` ) afiicher les trois premières lignes de `data`(resp les trois dérnières lignes de `data`)

In [7]:
data.head(3)

clump_thickness  uniformity_of_cell_size  uniformity_of_cell_shape  \
0                5                        1                         1   
1                5                        4                         4   
2                3                        1                         1   

   marginal_adhesion  single_epithelial_cell_size  bare_nuclei  \
0                  1                            2            1   
1                  5                            7           10   
2                  1                            2            2   

   bland_chromatin  normal_nucleoli  mitosis  class  
0                3                1        1      2  
1                3                2        1      2  
2                3                1        1      2

#### Question 4: Dans un seul fichier afficher les statistiques nécessaires pour `data` 

In [8]:
data.describe()

clump_thickness  uniformity_of_cell_size  uniformity_of_cell_shape  \
count       569.000000               569.000000                569.000000   
mean          4.539543                 3.184534                  3.265378   
std           2.896501                 3.002236                  2.955935   
min           1.000000                 1.000000                  1.000000   
25%           2.000000                 1.000000                  1.000000   
50%           4.000000                 1.000000                  2.000000   
75%           6.000000                 5.000000                  5.000000   
max          10.000000                10.000000                 10.000000   

       marginal_adhesion  single_epithelial_cell_size    bare_nuclei  \
count         569.000000                   569.000000     569.000000   
mean            2.845343                     3.298770   -2632.518453   
std             2.873626                     2.304775   16035.653408   
min             1.000000                     1.000000 -100000.000000   
25%             1.000000                     2.000000       1.000000   
50%             1.000000                     2.000000       1.000000   
75%             4.000000                     4.000000       8.000000   
max            10.000000                    10.000000      10.000000   

       bland_chromatin  normal_nucleoli     mitosis       class  
count       569.000000       569.000000  569.000000  569.000000  
mean          3.490334         2.989455    1.637961    2.731107  
std           2.324925         3.091315    1.773941    0.964018  
min           1.000000         1.000000    1.000000    2.000000  
25%           2.000000         1.000000    1.000000    2.000000  
50%           3.000000         1.000000    1.000000    2.000000  
75%           5.000000         4.000000    1.000000    4.000000  
max          10.000000        10.000000   10.000000    4.000000

#### Question 5:Extraire de `data` : les observations dans une variable`X` et les classes dans une variable `y`


In [9]:
X = data.drop('class', axis=1)
y = data['class']

#### Question 6: Diviser l'ensemble des observations X et l'ensemble de classes y, chacun en deux sous-ensembles :
- un sous-ensemble d'apprentissage : 80% de l'ensemble initial
- un sous-ensemble de test : 20% de l'ensemble initial
##### __Indication__ : Fixer à 0 le générateur aléatoire

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

<h1 align="center"><font size="7">K Nearest Neighbor(KNN)</font></h1>

In [11]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

#### En utilisant la méthode GridSearchCV, identifier le meilleur nombre de voisin (tester les valeurs de $n_{neighbors}\in \{1,2,3...25\}$, ainsi que la meilleur distance à utiliser (tester les valeurs de $p\in \{1,2,3,4,5,6 \}$

#### Afficher les meilleures valeurs des hyperparamètres réglés

#### Créer une instance __`final_model_knn`__ de l'algorithme de classification par KNN en utilisant les meilleurs hyperparamètres trouvés dans la question précédente
#### Entrainer __`final_model_knn`__ sur le sous-ensemble de données approprié.

#### Tracer la matrice de confusion de __`final_model_knn`__

### Afficher le __`classification_report`__ de __`final_model_knn`__

# Pour les autres algoritmes vous allez suivre pratiquement le meme raisonnement.

<h1 align="center"><font size="7">Decision tree - SALIM MEJDOUB</font></h1>

#### Importer le modèle __`DecisionTreeClassifier`__ à partir du module __`tree`__ du module __`sklearn`__.

#### Créer un dictionaire __`param_grid`__, en vue de configurer les trois hyperparamètres suivants :
##### - __`criterion`__ : qui peut être soit l'indice de Gini soit l'entropie
##### - __`max_depth`__ : qui varie de 1 à 19
##### - __`splitter`__ : qui peut être soit best ou random

#### Créer une instance de recherche par grille appliquée sur l'algorithme de classification par arbre de décision. La recherche doit tester toutes les combinaisons des valeurs des hyperparamètres de la grille de la question précédente. La recherche doit se faire à travers une validation croisée à 5 découpes.

#### Afficher les meilleures valeurs des hyperparamètres réglés

#### Créer une instance __`final_model_dt`__ de l'algorithme de classification par arbres de décision en utilisant les meilleurs hyperparamètres trouvés dans la question précédente

#### Afficher l'arbre de décision

In [20]:
features = ['clump_thickness', 'uniformity_of_cell_size', 'uniformity_of_cell_shape', 'marginal_adhesion',
       'single_epithelial_cell_size', 'bare_nuclei', 'bland_chromatin','normal_nucleoli', 'mitosis']

#### Afficher l'arbre sous forme de régles de décision

#### Tracer la matrice de confusion de __`final_model_dt`__

### Afficher le __`classification_report`__ de __`final_model_dt`__

<h1 align="center"><font size="7">LogisticRegression</font></h1> 

# Vous allez suivre pratiquement le même raisonnement. vous allez répondre aux même questions déjà posées

In [1]:
# Chargement de LogisticRegression
from sklearn.linear_model import LogisticRegression

#### Tracer la matrice de confusion de __`....`__

### Afficher le __`classification_report`__ de __`....`__

<h1 align="center"><font size="7">GaussianNB (Gaussian Naive Bayes)</font></h1>

# Vous allez suivre pratiquement le même raisonnement. vous allez répondre aux même questions déjà posées

In [25]:
# Chargement de GaussianNB
from sklearn.naive_bayes import GaussianNB

#### Créer une instance __`model`__ de l'algorithme de classification par GaussianNB.
#### Entrainer __`model`__ sur le sous-ensemble de données approprié.

#### Tracer la matrice de confusion de __`model`__

### Afficher le __`classification_report`__ de __`model`__

<h1 align="center"><font size="7"> Support vector machine (SVM)</font></h1>

# Vous allez suivre pratiquement le même raisonnement. vous allez répondre aux même questions déjà posées

In [12]:
from sklearn.svm import SVC

In [13]:
svclassifier = SVC(kernel='linear')
svclassifier.fit(X_train, y_train)

SVC(kernel='linear')

In [14]:
y_pred = svclassifier.predict(X_test)

In [15]:
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test,y_pred))

[[60  5]
 [ 2 47]]


In [16]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           2       0.97      0.92      0.94        65
           4       0.90      0.96      0.93        49

    accuracy                           0.94       114
   macro avg       0.94      0.94      0.94       114
weighted avg       0.94      0.94      0.94       114



# Etude compartive

| Algorithm          | weighted_avg_recall | weighted_avg_precsion | weighted_avg_F1-score|
|--------------------|---------------------|-----------------------|----------------------|
| KNN                | ?                   | ?                     | ?                    |
| Decision Tree      | ?                   | ?                     | ?                    |
| GaussianNB         | ?                   | ?                     | ?                    |
| LogisticRegression | ?                   | ?                     | ?                    |
| SVM                | 0.94                | 0.94                  | 0.94                 |

# Conclusion